rand_distr/src/geometric.rs

```diff
@@ -0,0 +1,119 @@
+//! The geometric distribution.
+
+use crate::{Distribution, Exp, Exp1};
+use num_traits::Float;
+use rand::Rng;
+use std::fmt;
+
+/// The geometric distribution `Geo(p)`.
+/// 
+/// This is the probability distribution of the number of failures before the
+/// first success in a series of Bernoulli trials. It has the density function
+/// `f(k) = (1 - p)^k p` for `k >= 0`, where `p` is the probability of success
+/// on each trial.
+/// 
+/// This is the discrete analogue of the [exponential distribution](crate::Exp).
+///
+/// # Example
+///
+/// ```
+/// use rand_distr::{Geo, Distribution};
+///
+/// let geo = Geo::new(0.5).unwrap();
+/// let x = geo.sample(&mut rand::thread_rng());
+/// println!("{} is from a Geo(0.5) distribution", x);
+/// ```
+#[derive(Copy, Clone, Debug)]
+pub struct Geo<F>
+where F: Float, Exp1: Distribution<F>
+{
+    exp: Exp<F>
+}
+
+/// Error type returned from `Geo::new`.
+#[derive(Clone, Copy, Debug, PartialEq, Eq)]
+pub enum Error {
+    /// `p < 0 || p > 1` or `nan`
+    InvalidProbability,
+}
+
+impl fmt::Display for Error {
+    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
+        f.write_str(match self {
+            Error::InvalidProbability => "p is NaN or outside the interval [0, 1] in geometric distribution",
+        })
+    }
+}
+
+#[cfg(feature = "std")]
+impl std::error::Error for Error {}
+
+impl<F> Geo<F>
+where F: Float, Exp1: Distribution<F>
+{
+    /// Construct a new `Geo` with the given shape parameter `p`
+    /// (probablity of success on each trial).
+    pub fn new(p: F) -> Result<Self, Error> {
+        let lambda = -F::ln(F::one() - p);
+        match Exp::new(lambda) {
+            Ok(exp) => Ok(Geo { exp }),
+            Err(_) => Err(Error::InvalidProbability)
+        }
+    }
+}
+
+impl<F> Distribution<u64> for Geo<F>
+where F: Float, Exp1: Distribution<F>
+{
+    fn sample<R: Rng + ?Sized>(&self, rng: &mut R) -> u64 {
+        let f = self.exp.sample(rng).floor();
+        num_traits::NumCast::from(f).unwrap_or(u64::max_value())
```

Comment:
This makes me wonder whether it would make more sense to return `f64`. On the other hand, we don't do that for `Binomial`.

Reply:
Yeah, I don't think that's the way to go for discrete distributions. As implemented, we don't have 100% coverage of the range `[1 << 53, u64::MAX]`, which the algorithm suggested by @peteroupc would provide.

I don't think lacking bitwise sampling is going to be a blocking issue in implementing it; as I understand it, `rng.gen::<u64>() & ((1 << k) - 1)` would work, and the benefits of bitwise sampling, namely fewer acceptance/rejection iterations, could then be gained by just using a bitmask.